examples/sriov.c

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+/*
+ * This file is part of libvfn.
+ *
+ * Copyright (C) 2022 The libvfn Authors. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the Free
+ * Software Foundation; either version 2 of the License, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ */
+
+#include <vfn/nvme.h>
+#include <vfn/pci.h>
+
+#include <nvme/types.h>
+
+#include "ccan/err/err.h"
+#include "ccan/opt/opt.h"
+#include "ccan/str/str.h"
+
+#include "common.h"
+
+static int vfnum;
+
+static struct opt_table opts[] = {
+	OPT_SUBTABLE(opts_base, NULL),
+	OPT_WITH_ARG("-s|--secondary VFNUM", opt_set_intval, opt_show_intval, &vfnum,
+		     "secondary controller virtual function number"),
+	OPT_ENDTABLE,
+};
+
+int main(int argc, char **argv)
+{
+	char *vf_bdf;
+	uint16_t scid;
+
+	struct nvme_ctrl ctrl = {}, sctrl = {};
+
+	opt_register_table(opts, NULL);
+	opt_parse(&argc, argv, opt_log_stderr_exit);
+
+	if (show_usage)
+		opt_usage_and_exit(NULL);
+
+	if (streq(bdf, ""))
+		opt_usage_exit_fail("missing --device parameter");
+
+	opt_free_table();
+
+	if (nvme_init(&ctrl, bdf, NULL))
+		err(1, "failed to init nvme controller");
+
+	vf_bdf = pci_get_vf_bdf(bdf, vfnum);
```

Comment:
It looks like this example assumes that VFs are already made on the system.

Don't we need to provide an API to actually enable the VFs from a PF `echo N > /sys/bus/pci/devices/.../sriov_numvfs` ?

Reply:
I'm a little conflicted. A helper to write `sriov_numvfs` could probably be useful, but I was thinking that we generally work with devices that are already configured on the system. We already rely on the user loading vfio-pci with `enable_sriov=1`.

Reply:
If needed, we can add a helper to initialize VFs later. For now, rely on the user doing that prior to starting the application.